presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PullUpExpressionInLambdaRules.java

```diff
@@ -295,12 +297,13 @@ private RowExpression getArgumentForRegexTypeExpression(RowExpression rowExpress
         @Override
         public Boolean visitSpecialForm(SpecialFormExpression specialForm, Boolean context)
         {
-            if (UNSUPPORTED_TYPES.contains(specialForm.getForm())) {
+            if (UNSUPPORTED_TYPES.contains(specialForm.getForm()) || specialForm.getArguments().size() > MAX_NUM_OF_ARGUMENTS) {
                 return false;
             }
-            List<RowExpression> validArguments = getValidArguments(specialForm);
-            Map<RowExpression, Boolean> validRowExpressionMap = specialForm.getArguments().stream().distinct().collect(toImmutableMap(identity(), x -> validArguments.contains(x) ? x.accept(this, context) : false));
```

Comment:
This is the problem, where we call `validArguments.contains(x)` for every argument. When validArguments is a list, it will be a O(N) operation for the contains, hence this line will be O(N^2). 
I verified by changing the List to Set, the latency of this optimizer is reduced significantly. 
[Using list](https://www.internalfb.com/intern/presto/query/?query_id=20240319_211452_00011_frxfm#runtime_stats) vs. [Using set ](https://www.internalfb.com/intern/presto/query/?query_id=20240319_211109_00005_frxfm#runtime_stats)

Reply:
Add a synthetic test with 100k IN list to verify this?

Reply:
In addition to change the List to Set, I also skip check when the number of arguments exceed 1000. This makes the synthetic test not working